manualintegrate: Add powsimp_rule for simplifying exponent expressions

salastro opened the pull request:
## Added powsimp_rule to simplify exponent expressions in integration

### Summary
Added a new integration rule, `powsimp_rule`, which simplifies exponent expressions by combining exponential terms where applicable. This is useful for cases like the ErfRule, where terms of the form `exp(a*x**2) * exp(b*x)` can be rewritten as `exp((a*x**2 + b*x))`.

### Changes
- Introduced `powsimp_rule` using `sympy.simplify.powsimp`.
- Integrated `powsimp_rule` into the manual integration strategy.
- Updated `test_manual.py` to include a test case validating the rule.

This enhancement improves integration handling by allowing more cases to be rewritten into a simpler form before applying further rules.

### Example
```python
In [1]: integrate(exp(-x**2)*exp(x))

Out [2]:
    1/4
√π⋅ℯ   ⋅erf(x - 1/2)
──────────────────
         2
```

### References to other Issues or PRs
Related to #27586. To be extended.

### Brief description of what is fixed or changed
This PR improves symbolic integration by adding a strategy that simplifies exponent expressions before applying further integration rules. This leads to more efficient computations and ensures better handling of cases involving exponentiated terms.

### Other comments
This is meant to be coupled with adding `erf(x)` to the liate integration by parts rule to solve special integrals of the form `erf(az)*exp(bz)`. I will open a pull request with the latter.

### Release Notes
<!-- BEGIN RELEASE NOTES -->
* integrals
  * Added `powsimp_rule` to simplify exponent expressions in manual integration.
<!-- END RELEASE NOTES -->

salastro commented:
I don't know why the authors test does not pass. I am sure I added my line to `.mailmap`. How to fix this?
```
SalahDin Rezk <s-salahdin.rezk@zewailcity.edu.eg> salastro <salah2112004@gmail.com>
```

vprayag2005 commented:
i think it is due to email issue
try this
```
SalahDin Rezk <salah2112004@gmail.com> salastro <salah2112004@gmail.com>
```

salastro commented:
> i think it is due to email issue try this
> 
> ```
> SalahDin Rezk <salah2112004@gmail.com> salastro <salah2112004@gmail.com>
> ```

thanks it worked

oscarbenjamin commented:
> > i think it is due to email issue try this
> > ```
> > SalahDin Rezk <salah2112004@gmail.com> salastro <salah2112004@gmail.com>
> > ```
> 
> thanks it worked

Which email address do you actually want to have recorded in the AUTHORS file?

oscarbenjamin commented on sympy/integrals/manualintegrate.py:
```diff
@@ -971,6 +972,27 @@ def exp_rule(integral):
+
+    simplified = powsimp(integrand, combine='exp')
+
+    return None if simplified == integrand else RewriteRule(integrand, symbol, simplified, integral_steps(simplified, symbol))
```
This line is long. There is no need for it to be long though:
```python
if simplified != integrand:
    steps = integral_steps(simplified, symbol)
    return RewriteRule(integrand, symbol, simplified, steps)
```

salastro replied:
Done

salastro commented:
There is a fail in the Laplace Transform tests:

```diff
  -     b*LaplaceTransform(exp(2*t**2), t, s),
  +     b*(oo*sign(exp(-s**2/8)) + sqrt(2)*sqrt(pi)*exp(-s**2/8)*erfi(sqrt(2)*s/4)/4),
```

I think combining exponential terms now gives a solution to previously unsolved LTs. I tried Wolfram Alpha but it did not give me this expression and said "No result."

salastro commented:
> > > i think it is due to email issue try this
> > > ```
> > > SalahDin Rezk <salah2112004@gmail.com> salastro <salah2112004@gmail.com>
> > > ```
> > 
> > 
> > thanks it worked
> 
> Which email address do you actually want to have recorded in the AUTHORS file?

I originally wanted my university email s-salahdin.rezk@zewailcity.edu.eg.

salastro commented:
I am not sure why the code quality test is not passing. Locally, it gives me:

```
[I] (venv) salastro@LenovoIdeapadGaming3 ~/.s/sympy (powsimp) [0|1]> ./bin/test quality
                                                                     flake8 sympy/
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.13.2, pytest-8.3.5, pluggy-1.5.0
architecture: 64-bit
cache:        yes
ground types: python

rootdir: /home/salastro/.srcpkgs/sympy
configfile: pyproject.toml
plugins: timeout-2.3.1, split-0.10.0, xdist-3.6.1, hypothesis-6.130.4, doctestplus-1.4.0
collected 7 items

sympy/testing/tests/test_code_quality.py .......                                                                                                                                        [100%]

===================================================================================== 7 passed in 11.02s ======================================================================================
```

oscarbenjamin commented on .mailmap:
```diff
@@ -1331,6 +1331,7 @@ Saket Kumar Singh <saketkumar1202@gmail.com>
 Saketh <alurusaisaketh@gmail.com>
 Sakirul Alam <binarysakir@gmail.com>
 Saksham Alok <sakshamalok13@gmail.com> Saksham-13 <sakshamalok13@gmail.com>
+SalahDin Rezk <salah2112004@gmail.com> salastro <salah2112004@gmail.com>
```
> I originally wanted my university email [s-salahdin.rezk@zewailcity.edu.eg](mailto:s-salahdin.rezk@zewailcity.edu.eg).

Change this to:
```
SalahDin Rezk <s-salahdin.rezk@zewailcity.edu.eg> SalahDin Rezk <salah2112004@gmail.com>
SalahDin Rezk <s-salahdin.rezk@zewailcity.edu.eg> salastro <salah2112004@gmail.com>
```

oscarbenjamin commented:
> I am not sure why the code quality test is not passing.

It is because of a new release of flake8. To reproduce locally:
```
pip install -U flake8
flake8 sympy
```
See gh-27855.

salastro commented:
I tested `LT[2*t**2]` on Mathematica and it yielded the following results. I think there is no problem then and the Laplace test case should be changed.
[Laplace.pdf](https://github.com/user-attachments/files/19532164/Laplace.pdf)

salastro commented:
Fixed the laplace test issue. Anything else to do before merging?

salastro commented:
Fixed the laplace test issue. Anything else to do before merging?

salastro committed:
manualintegrate: Add powsimp_rule for simplifying exponent expressions

Added a new integration rule, powsimp_rule, which simplifies exponent
expressions by combining exponential terms where applicable. This is
useful for cases like the ErfRule, where terms of the form
exp(a*x**2) * exp(b*x) can be rewritten as exp((a*x**2 + b*x)).

- Introduced powsimp_rule using sympy.simplify.powsimp.
- Integrated powsimp_rule into the manual integration strategy.
- Updated test_manual.py to include a test case validating the rule.

This enhancement improves integration handling by allowing more cases
to be rewritten into a simpler form before applying further rules.

Example:

In [1]: integrate(exp(-x**2)*exp(x))

Out [2]:
    1/4
√π⋅ℯ   ⋅erf(x - 1/2)
────────────────────
         2

salastro commented:
Fixed failing tests. This should be good to merge.

oscarbenjamin commented on sympy/integrals/tests/test_laplace.py:
```diff
@@ -94,7 +94,7 @@ def test_laplace_transform():
     assert (LT(exp(-2*t**2), t, s) ==
             (sqrt(2)*sqrt(pi)*exp(s**2/8)*erfc(sqrt(2)*s/4)/4, 0, True))
     assert (LT(b*exp(2*t**2), t, s) ==
-            (b*LaplaceTransform(exp(2*t**2), t, s), -oo, True))
+            (b * (oo * sign(exp(-s**2 / 8)) + (sqrt(2) * sqrt(pi) * exp(-s**2 / 8) * erfi(s * sqrt(2) / 4)) / 4), -oo, True))
```
Why is there an infinity here? I think this should remain unevaluated.

salastro replied:
The closed form of LT exp(2*t**2) is:

![image](https://github.com/user-attachments/assets/c7c84952-9a40-4c2d-b630-1f5c065f075e)

I don't know why did an infinity get added instead of the correct term.

salastro replied:
There should be a way to fix it but I am not deep inside the laplace transform code.

salastro replied:
It should be caused by the laplace integral definition getting simplified into one exponential and then solved as an erf term.

salastro replied:
Okay so I believe that Mathematica's output is faulty bc this LT does not actually converge in any possible way. I couldn't find a way to show that the integral diverges, so I put the condition on the laplace transform function to check whether the final answer has an infinity or not, and if yes, it assumes that the form does not converge.

salastro commented:
The closed form of LT exp(2*t**2) is:

![image](https://github.com/user-attachments/assets/c7c84952-9a40-4c2d-b630-1f5c065f075e)

I don't know why did an infinity get added instead of the correct term.

salastro committed:
manualintegrate: move powsimp_rule into alternatives node

All rewrite rules should be in the `alternatives` node to continue
swtiching between different rewrite rules, reusing them, and reaching a
known rule in the lookup integration table.

salastro commented:
I moved the `powsimp_rule` into the aternatives node. However, I am not sure if its order in the rules hierarchy optimizes results or not.

salastro commented on sympy/integrals/laplace.py:
```diff
@@ -1251,6 +1251,11 @@ def _laplace_transform(fn, t_, s_, *, simplify):
             else:
                 r = (LaplaceTransform(ft, t_, s_), S.NegativeInfinity, True)
         (ri_, pi_, ci_) = r
+        if S(ri_).has(S.Infinity, S.NegativeInfinity, S.ComplexInfinity):
+            _debug(f"Result for term {ft} contains infinity, returning k*Unevaluated.")
```
It is caused by the integration limits, so it is related to the evaluation function. The indefinite integral of `exp(-s*t)*exp(2*t**2)` is accurate i.e. `sqrt(2)*sqrt(pi)*exp(-s**2/8)*erfi(sqrt(2)*(-s + 4*t)/4)/4`, but the issue arises when infinity is substituted for t. The expression is supposed to diverge to infinity, yet the evaluator outputs the weird `oo*sign(exp(-s**2/*))` term. I believe that Mathematica may be wrong in this case because I tested substituting infinity into the integral expression and it gave me a similar term:

![image](https://github.com/user-attachments/assets/a3b61969-fecb-4a88-bc72-5b72d2ff0e5c)

![image](https://github.com/user-attachments/assets/da329dae-4e24-4bfd-a444-34b7892bb2be)

salastro replied:
I don't have any other approach in mind. Should I instead try to tackle the evaluation step?

salastro replied:
I don't have any other approach in mind. Should I instead try to tackle the evaluation step?

asmeurer commented:
The `integrate(exp(2*t**2)*exp(-t*z), (t, 0, oo))` issue is definitely coming from a misapplication of the FTC, which is unrelated to the changes here. 

The computed antiderivative is correct here. The problem is that the limit in `_eval_interval` return `oo*sign(exp(-z**2/8))` and there's no code there to handle this case. I guess it normally assumes that if one of the limits is oo that the infinity will just absorb everything, but that doesn't happen in this case because of the symbolic parameters. It probably wouldn't hurt to be a little more conservative in _eval_interval and avoid returning things when one of the limits is infinite to avoid returning wrong or nonsense results. OTOH the result here isn't completely bad since it is a nonsense expression for something that doesn't actually exist.

asmeurer commented:
Ah, so I see now from the comments that this is affecting an actual test failure. I agree that the changes here to laplace.py are not good. If we want to fix this, we should consider adding some more special handling to _eval_interval, as I described above. I'm not sure if it would break other things, but a check [here](https://github.com/sympy/sympy/blob/089edf0ea752497120cbaa1a5af4b5d4bcab9a85/sympy/core/expr.py#L986) would help. Maybe a check for `if A.has(oo) and not B.is_finite:`. Actually, unless I'm misreading the code, I don't even see a clear way for `_eval_interval` to "fail", other than raising NotImplementedError (typically from `limit`), so that should probably also be fixed.

oscarbenjamin commented:
I think probably `_eval_interval` and related code should all be rewritten perhaps from scratch. First thing to do is just to move the code into one place rather than spreading it all over the place. There is no reason that `_eval_interval` should be an `Expr` method rather than a function in `integrals.py`. Likewise all the Piecewise stuff should be moved to `integrals.py` as well.

salastro commented:
> Ah, so I see now from the comments that this is affecting an actual test failure. I agree that the changes here to laplace.py are not good. If we want to fix this, we should consider adding some more special handling to _eval_interval, as I described above. I'm not sure if it would break other things, but a check [here](https://github.com/sympy/sympy/blob/089edf0ea752497120cbaa1a5af4b5d4bcab9a85/sympy/core/expr.py#L986) would help. Maybe a check for `if A.has(oo) and not B.is_finite:`. Actually, unless I'm misreading the code, I don't even see a clear way for `_eval_interval` to "fail", other than raising NotImplementedError (typically from `limit`), so that should probably also be fixed.

is adding 

```python
        if A.has(S.Infinity, S.NegativeInfinity, S.ComplexInfinity) or B.has(S.Infinity, S.NegativeInfinity, S.ComplexInfinity):
            raise ValueError("Interval appears to diverge at an endpoint.")

        value = B - A
```
or
```python
                if C.has(S.Infinity, S.NegativeInfinity, S.ComplexInfinity):
                    raise ValueError(f"Integral diverges at endpoint {c}")
```

sufficient? Should I fix the internal code of the limit or skip it for now?

oscarbenjamin commented:
I don't think it is necessary to fix the limit handling code but the changes in laplace.py should be reverted. The failing test would need to be changed I guess.

salastro commented:
How should the behaviour be given the failed tests? I am a bit unsure.

oscarbenjamin commented:
The failures all seem to be caused by the check that you have added in `_eval_endpoint`.

salastro commented:
Shold I instead return `S.NaN`?

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -990,6 +990,15 @@ def _eval_endpoint(left):
 
         value = B - A
 
+        # Catch undefined forms like oo - oo or oo * symbolic
+        if isinstance(value, Add) and any(arg.is_infinite for arg in value.args):
+            if any(a.is_infinite for a in value.args) and any(a.has(x) for a in value.args):
+                return S.NaN
+
+        if isinstance(value, Mul) and any(arg.is_infinite for arg in value.args):
+            if any(arg.has(x) for arg in value.args if arg != S.Infinity and arg != S.NegativeInfinity):
+                return S.NaN
```
I don't understand why you have added this. Why not just remove it?

It doesn't really fix anything in a robust way.

salastro replied:
Okay.

salastro committed:
Mark invalid Laplace transform test as xfail

oscarbenjamin commented on sympy/integrals/tests/test_laplace.py:
```diff
@@ -488,6 +492,16 @@ def test_laplace_transform():
     assert LT(Mt, t, s, legacy_matrix=False, noconds=True) == Ms
 
 
+@slow
+@XFAIL
+def test_laplace_transform_exp_2t2():
+    LT = laplace_transform
+    b = symbols('b', positive=True)
+    t = symbols('t')
+    assert (LT(b*exp(2*t**2), t, s) ==
+            (b*LaplaceTransform(exp(2*t**2), t, s), -oo, True))
```
There should be a comment here that the test belongs in `test_laplace_transform` once fixed and is already there but commented out.

For an XFAIL test I think it is better to assert that the output is not whatever it returns like:
```
In [9]: wrong = (b*(oo*sign(exp(-s**2/8)) + sqrt(2)*sqrt(pi)*exp(-s**2/8)*erfi(sqrt(2)*s/4)/4), -oo, True)

In [10]: assert LT(b*exp(2*t**2), t, s) != wrong
---------------------------------------------------------------------------
AssertionError
```
Then if the output ever changes to something else we will know.
